walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,158 @@
+            }
+            currentStart =  tr.end();
+        }
+        TimeRange newTimeRange = TimeRange.fromStartEnd(currentStart,1440,false);
+        if(newTimeRange.start() != newTimeRange.end() && newTimeRange.duration() >= request.getDuration()) {
+            ret.add(newTimeRange);
+        } 
+        if(request.getOptionalAttendees().size() > 0) {
+            ret = tryToFitOptionalAttendees(ret,events,request);
+        }
+        return ret;
+  }
+
+  /**
+    * Function that determines whether any attendees of the requested meeting have any other meeting that day
+    * @param events a collection of events
+    * @param request a meeting request 
+    * @param optional a boolean flag that determines whether the function checks optional attendees or mandatory attendees
+    * @return ret an ARraylist of the other Events
+    */
+  public ArrayList<Event> doAttendeesHaveAnyOtherMeetings(Collection<Event> events, MeetingRequest request, boolean optional) {
+      ArrayList<Event> ret = new ArrayList<Event>();
+      for(Event e : events){
+        if(doAttendeesOverlap(e, request, optional)){
+            ret.add(e);
+        }
+      }
+      return ret;
+  }
+
+  /**
+    * Function that checks a single event and determines whether the requested meetings attendees are included in it
+    * @param event the Event that the MeetingRequest will be checked againt
+    * @param request the requested Meeting
+    * @return a boolean that respresents whether the attendees overlap
+    */
+  public boolean doAttendeesOverlap(Event event, MeetingRequest request, boolean optional) {
+      for(String eventAttendee : event.getAttendees()) {
+          for(String requestAttendee : optional ? request.getOptionalAttendees() : request.getAttendees()) {
+              if(eventAttendee.compareTo(requestAttendee) == 0) {
+                  return true;
+              }
+          }
+      }
+      return false;
+  }
+  /**
+    * Function that iterates through the timeranges and removes and overlapping timeranges
+    * @param timeranges an Arraylist of TimeRanges
+    * @return void 
+    */
+  public void removeOverlap(ArrayList<TimeRange> timeranges){ 
+      for(int i = 1; i < timeranges.size(); ++i){
+        if(timeranges.get(i-1).overlaps(timeranges.get(i))) {
+            TimeRange newrange = mergeTimeRanges(timeranges.get(i-1), timeranges.get(i));
+            timeranges.set(i-1,newrange);
+            timeranges.remove(i);
+        }
+      }
+  }
+  /**
+    * Function that merges two overalapping timeranges
+    * @param tr1 a TimeRange that needs to be merged
+    * @param tr2 another TimeRange that needs to be merged
+    * @return new TimeRange created from the overlap between tr1 and tr2
+   */
+  public TimeRange mergeTimeRanges(TimeRange tr1, TimeRange tr2) {
+      int earliest;
+      int latest;
+      if(tr1.start() < tr2.start()) {
+        earliest = tr1.start();
+      } else {
+        earliest = tr2.start();
+      }
```

Comment:
earliest = Math.min(tr1.start(), tr2.start());

same for latest.

Also, where is the check to ensure that the timelines do overlap?

Reply:
That is taken care of in the previous function